build,refactor: update development tooling and package configuration (#614)
C2SM/icon4py: README.md

# ICON4Py

This repository hosts a work-in-progress Python implementation of the ICON climate and weather model. Additionally, it includes `icon4pytools`, a collection of command-line interfaces (CLIs), and utilities required for the integration of ICON4Py code into the ICON Fortran model. ICON4Py leverages [GT4Py](https://github.com/GridTools/gt4py) to ensure efficient and performance portable implementations of these components.

## Project Structure

The repository is organized as a _monorepo_, where various ICON model components and utilities are developed as independent Python namespace packages in subfolders. An `icon4py` Python package is defined at the root folder just to collect the specific versions of the different components as project dependencies. The component packages can also be installed independently, although since they are not (yet) available from a package repository, they need to be installed from their specific location within this repository. This can be easily done by following the [Installation instructions](#installation-instructions) explained below.

## License

ICON4Py is licensed under the terms of the BSD-3-Clause.

## Installation instructions

> Since this project is still in a highly experimental state, it is not yet available as a regular Python distribution project through PyPI. The expected installation procedure is to clone the [https://github.com/C2SM/icon4py](https://github.com/C2SM/icon4py) repository and install it in a venv using the following development workflow.  

Reviewer: ```suggestion
Since this project is still in a highly experimental state, it is not yet available as a regular Python distribution project through PyPI. The installation procedure comprises cloning the [https://github.com/C2SM/icon4py](https://github.com/C2SM/icon4py) repository and install it in a `venv` using the following development workflow.  
```

### System dependencies

ICON4Py requires **_Python >= 3.10_** and **_boost >= 1.85.0_**, and uses the `uv` tool to manage the development workflow. `uv` is a versatile tool which bundles together functionality from different applications: it can work as a _fast_ Python package manager (like `pip`), as a dependency version exporter (like `pip-tools`), as a Python application runner (like `pipx`) or as a full project development manager (like `hatch`). 
`uv` can be installed in different ways (check its [installation instructions](https://docs.astral.sh/uv/getting-started/installation/)), like using the recommended standalone installer:

```bash
$ curl -LsSf https://astral.sh/uv/install.sh | sh 
```

> although it can also be installed with `pip` / `pipx`

Reviewer 2: would delete this unless there is a reason why you would ever need this fallback

```bash
$ pipx install uv
```

### ICON4Py development environment

Once `uv` is installed in your system, it is enough to clone this repository and let `uv` handling the installation of the development environment.

```bash
# Clone the repository
git clone git@github.com:C2SM/icon4py.git
cd icon4py

# Let uv create the development environment at `.venv`.
# The `--extra all` option tells uv to install all optional
# dependencies of icon4py so it is not strictly necessary
uv sync --extra all

# Activate the virtual environment and start writing code!
source .venv/bin/activate
```

> The new `venv` is a standard Python virtual environment with all the necessary run-time and development dependencies, where all the icon4py subpackages have been already installed in editable mode. New packages can be installed with the `uv pip` subcommand which emulates the `pip` interface or even using the regular `pip` which is also available in the venv (although it's usually much slower and not really recommended).   

Reviewer: ```suggestion
The new `venv` is a standard Python virtual environment with all the necessary run-time and development dependencies, where pre-installed icon4py subpackages in editable mode. New packages can be installed with the `uv pip` subcommand which emulates the `pip` interface or even using the regular `pip` which is also available in the venv (although it's usually much slower and not really recommended).   
```

Author: I slightly changed the whole paragraph and split it into two. Check the new version.

pytest -v path/to/test/folder
```

> Nonetheless, we also recommended to use `nox` to run the parametrized test suite:

Reviewer 2: It's unclear why we recommend this

> Nonetheless, we also recommended to use `nox` to run the parametrized test suite:

Reviewer 2: ```suggestion
Nonetheless, we also recommend to use `nox` to run the parametrized test suite:
```

```bash
# List all available test sessions (colored items are the default sessions)
nox -l

# Run all parametrized cases of a session
nox -s 'test_common'

# Run a test session for a specific python version and parameter values
nox -s 'test_atmosphere_advection-3.10(datatest=True)'
```

> The default `tox` environment is configured to generate HTML test coverage reports in `_reports/coverage_html/`.

Reviewer 2: does tox still exist?

Author: No, the whole sentence is wrong.
